flatdata-cpp/src/SchemaUtils.re2c.cpp

```diff
@@ -0,0 +1,231 @@
+        case Token::KIND::PUNCTUATION:
+        case Token::KIND::INVALID:
+            return t1.character == t2.character;
+        case Token::KIND::NUMBER:
+            return t1.number == t2.number;
+        case Token::KIND::END:
+            return true;
+        }
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool
+operator!=( const Token& t1, const Token& t2 )
+{
+    return !( t1 == t2 );
+}
+
+class Lexer
+{
+public:
+    Lexer( const char* data, size_t size )
+        : m_data( data )
+        , m_end( data + size )
+    {
+    }
+
+    Token next_token( );
+
+private:
+    const char* m_data;
+    const char* m_end;
+};
+
+Token
+Lexer::next_token( )
+{
+    /*!types:re2c*/
+
+    const char* marker;
+    const char* context_marker;
+    YYCONDTYPE c = yycinit;
+
+    for ( ;; )
+    {
+        if ( m_data == m_end )
+        {
+            Token t;
+            t.kind = Token::KIND::END;
+            return t;
+        }
+
+        const char* start = m_data;
+        // enough to hold UINT64_MAX as decimal or hex + nul byte
+        char number_buffer[ 21 ];
+
+        // Using the custom interface to be able to deal inputs that are not nul
+        // terminated. The nul byte is still an invalid character so we can use
+        // it as a sentinel.
+#define YYCTYPE char
+#define YYPEEK( ) ( ( m_data < m_end ) ? *m_data : 0 )
+#define YYSKIP( ) ( ++m_data )
+#define YYBACKUP( ) ( marker = m_data )
+#define YYRESTORE( ) ( m_data = marker )
+#define YYBACKUPCTX( ) ( context_marker = m_data )
+#define YYRESTORECTX( ) ( m_data = context_marker )
+
+        // clang-format off
+        /*!re2c
+        // input is complete, no buffer filling
+        re2c:yyfill:enable = 0;
+        re2c:indent:string = "    ";
+        re2c:indent:top = 2;
+        re2c:define:YYGETCONDITION = "c";
+        re2c:define:YYGETCONDITION:naked = 1;
+        re2c:define:YYSETCONDITION = "c = @@;";
+        re2c:define:YYSETCONDITION:naked = 1;
+        re2c:cond:divider = "// condition: @@";
+
+        whitespace = [ \t\r\n];
+        nul = '\000';
+        word = "@"? [A-Za-z_] [A-Za-z0-9_]*;
+        decimal = [0-9]{1,20};
+        hex = "0x" [0-9a-fA-F]{1,19};
+        punctuation = [;:{}<>\[\]];
+
+        <*> * {
+            Token t;
+            t.kind = Token::KIND::INVALID;
+            t.character = yych;
+            return t;
+        }
+
+        <init> "//" :=> cpp_comment
+        <init> "/*" :=> c_comment
+
+        <init> whitespace { continue; }
+        <init> word
+        {
+            Token t;
+            t.kind = Token::KIND::WORD;
+            t.word.start = start;
+            t.word.end = m_data;
+            return t;
+        }
+        <init> punctuation
+        {
+            Token t;
+            t.kind = Token::KIND::PUNCTUATION;
+            t.character = yych;
+            return t;
+        }
+        <init> decimal / ( whitespace | punctuation | nul ) {
```

Comment:
This should not be required for ignoring whitespaces and comments, should it? And I am not sure that we need to ignore hex vs dec.
Also: negative numbers might be broken

Reply:
This prevents the parser from accepting "123abc" as <Number: 123> <Word: abc>.